packages/terra-form-select/src/native-select/NativeSelect.jsx

```diff
@@ -0,0 +1,278 @@
+  /**
+   * @private Callback function not intended for use with this API, but if set pass it through to the element regardless.
+   */
+  onBlur: PropTypes.func,
+  /**
+   * @private Callback function not intended for use with this API, but if set pass it through to the element regardless.
+   */
+  onFocus: PropTypes.func,
+  /**
+   * @private Callback function not intended for use with this API, but if set pass it through to the element regardless.
+   */
+  onMouseDown: PropTypes.func,
+};
+
+const defaultProps = {
+  disabled: false,
+  isIncomplete: false,
+  isInvalid: false,
+  options: [],
+  required: false,
+};
+
+const createPlaceholder = (placeholder, intl) => {
+  if (!placeholder) {
+    return undefined;
+  }
+  const display = placeholder.display ? placeholder.display : intl.formatMessage({ id: 'Terra.form.select.defaultDisplay' });
+  const value = placeholder.value ? placeholder.value : defaultPlaceholderValue;
+  const attrs = placeholder.allowClear ? {} : { disabled: true, hidden: true };
+  return <option value={value} {...attrs}>{display}</option>;
+};
+
+const createOptions = options => {
+  const currentOptGroupKeys = [];
+
+  return options.map(current => {
+    if (current.options) {
+      const optGroupKeyIndex = getOptGroupKeyIndex(current.display, currentOptGroupKeys);
+      const optGroupKey = getOptGroupKey(current.display, optGroupKeyIndex);
+      currentOptGroupKeys.push(optGroupKey);
+
+      return (
+        <optgroup key={optGroupKey} label={current.display}>
+          {createOptions(current.options)}
+        </optgroup>
+      );
+    }
+    return <option key={`${current.value}`} value={current.value}>{current.display}</option>;
+  });
+};
+
+const NativeSelect = ({
+  ariaDescribedBy,
+  ariaLabel,
+  attrs,
+  disabled,
+  defaultValue,
+  id,
+  intl,
+  isInvalid,
+  isIncomplete,
+  onBlur,
+  onChange,
+  onFocus,
+  onMouseDown,
+  options,
+  placeholder,
+  refCallback,
+  required,
+  value,
+  ...customProps
+}) => {
+  const [uncontrolledValue, setUncontrolledValue] = useState(!isValuePresent(value) ? defaultValue || getFirstValue(options, placeholder) : undefined);
+  const refIsControlled = useRef(isValuePresent(value));
+  const refSelect = useRef();
+  const theme = React.useContext(ThemeContext);
+
+  const handleOnMouseDown = (event) => {
+    refSelect.current.setAttribute('data-focus-interaction', 'mouse');
+    if (onMouseDown) {
+      onMouseDown(event);
+    }
+  };
+  const handleOnBlur = (event) => {
+    refSelect.current.setAttribute('data-focus-interaction', 'none');
+    if (onBlur) {
+      onBlur(event);
+    }
+  };
+  const handleOnFocus = (event) => {
+    if (refSelect.current.getAttribute('data-focus-interaction') !== 'mouse') {
+      refSelect.current.setAttribute('data-focus-interaction', 'keyboard');
+    }
+    if (onFocus) {
+      onFocus(event);
+    }
+  };
+  const handleOnChange = event => {
```

Comment:
we should consider using useCallback for these funcitons :  https://reactjs.org/docs/hooks-reference.html#usecallback

Reply:
Looking at some the implementation guides it appears that useCallback's optimization may be more expensive than its benefit as the inline function is still generated. The ideal usage would be more if the implementation relied more heavily on memoization.
https://dmitripavlutin.com/dont-overuse-react-usecallback/
https://kentcdodds.com/blog/usememo-and-usecallback